File: backend/src/services/game-logic.service.ts
import { GameState } from '../models/game-state';
import { Card } from '../models/card';
import { Deck } from '../models/deck';
import { NodeDefinition } from '../models/node';
import { MapGeneratorService } from './map-generator.service';
import { EventSpawnerService } from './event-spawner.service';

export class GameLogicService {
  private mapGenerator = new MapGeneratorService();
  private eventSpawner = new EventSpawnerService();

  createInitialState(): GameState {
        const cards: Card[] = [
      {
        id: 'strike', name: 'Strike', cost: 1, type: 'attack',
        properties: { damage: 3 },
        effectId: 'fx-strike-normal', enhancedEffectId: 'fx-strike-enhanced',
        effect: { description: 'Deal 3 damage to one enemy.' },
        enhancedEffect: { description: 'Deal 5 damage to one enemy.' },
      },
      {
        id: 'shield', name: 'Shield', cost: 1, type: 'defense',
        properties: { block: 3 },
        effectId: 'fx-shield-normal', enhancedEffectId: 'fx-shield-enhanced',
        effect: { description: 'Gain 2 shield.' },
        enhancedEffect: { description: 'Gain 4 shield.' },
      },
      {
        id: 'focus', name: 'Focus', cost: 2, type: 'utility',
        properties: { manaGain: 2 },
      },
      {
        id: 'bash', name: 'Bash', cost: 2, type: 'attack',
        properties: { damage: 5 },
        effectId: 'fx-strike-enhanced', enhancedEffectId: 'fx-comp-strike-enhanced',
        effect: { description: 'Deal 5 damage to one enemy.' },
        enhancedEffect: { description: 'Deal 6 damage to one enemy.' },
      },
      {
        id: 'heal', name: 'Heal', cost: 1, type: 'defense',
        properties: { recover: 2 },
      },
      {
        id: 'charge', name: 'Charge', cost: 1, type: 'utility',
        properties: { speed: 1 },
      },
    ];

    const graph = this.mapGenerator.generate({ minNodes: 20, maxNodes: 24, minLayers: 5, maxLayers: 7 });
    // enforce event spawn rules on generated nodes
    graph.nodes = this.eventSpawner.assignEvents(graph.nodes);
    const deck: Deck = { cardIds: cards.map(card => card.id) };

    return {
      player: {
        id: 'player-1',
        life: 20,
        mana: 3,
        deck,
        hand: ['strike', 'shield', 'focus'],
        discard: [],
        position: 'start'
      },
      graph,
      cards,
      companions: [],
      history: ['New run created']
    };
  }

  movePlayer(state: GameState, nextNodeId: string): GameState {
    const isForwardMove = state.graph.edges.some(edge => edge.from === state.player.position && edge.to === nextNodeId);
    if (!isForwardMove) {
      return { ...state, history: [...state.history, `invalid move attempted to ${nextNodeId}`] };
    }

    return {
      ...state,
      player: { ...state.player, position: nextNodeId },
      history: [...state.history, `moved to ${nextNodeId}`]
    };
  }

  playCard(state: GameState, cardId: string): GameState {
    const card = state.cards.find(c => c.id === cardId);
    if (!card || !state.player.hand.includes(cardId) || card.cost > state.player.mana) {
      return { ...state, history: [...state.history, `could not play ${cardId}`] };
    }

    const remainingHand = state.player.hand.filter(id => id !== cardId);
    const discard = [...(state.player.discard ?? []), cardId];

    return {
      ...state,
      player: {
        ...state.player,
        mana: state.player.mana - card.cost,
        hand: remainingHand,
        discard
      },
      history: [...state.history, `played ${card.name}`]
    };
  }
}

File: backend-csharp/TheGameIsOn.API/Services/GameLogicService.cs
using System.Text.Json;
using TheGameIsOn.API.Models;

namespace TheGameIsOn.API.Services;

public interface IGameLogicService
{
    Task<GameState> CreateInitialStateAsync();
    GameState MovePlayer(GameState state, string nextNodeId);
    GameState PlayCard(GameState state, string cardId);
}

public class GameLogicService : IGameLogicService
{
    private readonly IMapGeneratorService _mapGenerator;
    private readonly IEventSpawnerService _eventSpawner;

    public GameLogicService(IMapGeneratorService mapGenerator, IEventSpawnerService eventSpawner)
    {
        _mapGenerator = mapGenerator;
        _eventSpawner = eventSpawner;
    }

    public async Task<GameState> CreateInitialStateAsync()
    {
        var cards = new List<Card>
        {
            new()
            {
                Id = "strike", Name = "Strike", Cost = 1, Type = CardType.Attack, Element = CardElement.Neutral,
                Properties = JsonProperties(("damage", 3)),
                EffectId = "fx-strike-normal", EnhancedEffectId = "fx-strike-enhanced",
                Effect = new CardEffectRef("Deal 3 damage to one enemy."),
                EnhancedEffect = new CardEffectRef("Deal 5 damage to one enemy.")
            },
            new()
            {
                Id = "shield", Name = "Shield", Cost = 1, Type = CardType.Defense, Element = CardElement.Neutral,
                Properties = JsonProperties(("block", 3)),
                EffectId = "fx-shield-normal", EnhancedEffectId = "fx-shield-enhanced",
                Effect = new CardEffectRef("Gain 2 shield."),
                EnhancedEffect = new CardEffectRef("Gain 4 shield.")
            },
            new()
            {
                Id = "focus", Name = "Focus", Cost = 2, Type = CardType.Utility, Element = CardElement.Neutral,
                Properties = JsonProperties(("manaGain", 2))
            },
            new()
            {
                Id = "bash", Name = "Bash", Cost = 2, Type = CardType.Attack, Element = CardElement.Neutral,
                Properties = JsonProperties(("damage", 5)),
                EffectId = "fx-strike-enhanced", EnhancedEffectId = "fx-comp-strike-enhanced",
                Effect = new CardEffectRef("Deal 5 damage to one enemy."),
                EnhancedEffect = new CardEffectRef("Deal 6 damage to one enemy.")
            },
            new()
            {
                Id = "heal", Name = "Heal", Cost = 1, Type = CardType.Defense, Element = CardElement.Neutral,
                Properties = JsonProperties(("recover", 2))
            },
            new()
            {
                Id = "charge", Name = "Charge", Cost = 1, Type = CardType.Utility, Element = CardElement.Neutral,
                Properties = JsonProperties(("speed", 1))
            }
        };

        var graph = _mapGenerator.Generate(new MapGenerationOptions
        {
            MinNodes = 20, MaxNodes = 24, MinLayers = 5, MaxLayers = 7
        });

        var nodes = await _eventSpawner.AssignEventsAsync(graph.Nodes);
        graph = graph with { Nodes = nodes };

        var deck = new Deck { CardIds = cards.Select(c => c.Id).ToList() };

        return new GameState
        {
            Player = new Player
            {
                Id = "player-1",
                Life = 20,
                Mana = 3,
                Gold = 0,
                Deck = deck,
                Hand = new List<string> { "strike", "shield", "focus" },
                Discard = new(),
                Position = "start",
                EncounterCount = 0
            },
            Graph = graph,
            Cards = cards,
            Companions = new(),
            History = new List<string> { "New run created" }
        };
    }

    public GameState MovePlayer(GameState state, string nextNodeId)
    {
        var isForwardMove = state.Graph.Edges.Any(e => e.From == state.Player.Position && e.To == nextNodeId);
        if (!isForwardMove)
        {
            return new GameState
            {
                Player = state.Player,
                Graph = state.Graph,
                Cards = state.Cards,
                Companions = state.Companions,
                History = new List<string>(state.History) { $"invalid move attempted to {nextNodeId}" },
                Battle = state.Battle,
                PendingAbilityChoices = state.PendingAbilityChoices
            };
        }

        return new GameState
        {
            Player = state.Player with { Position = nextNodeId },
            Graph = state.Graph,
            Cards = state.Cards,
            Companions = state.Companions,
            History = new List<string>(state.History) { $"moved to {nextNodeId}" },
            Battle = state.Battle,
            PendingAbilityChoices = state.PendingAbilityChoices
        };
    }

    public GameState PlayCard(GameState state, string cardId)
    {
        var card = state.Cards.FirstOrDefault(c => c.Id == cardId);
        if (card is null || !state.Player.Hand.Contains(cardId) || card.Cost > state.Player.Mana)
        {
            return new GameState
            {
                Player = state.Player,
                Graph = state.Graph,
                Cards = state.Cards,
                Companions = state.Companions,
                History = new List<string>(state.History) { $"could not play {cardId}" },
                Battle = state.Battle,
                PendingAbilityChoices = state.PendingAbilityChoices
            };
        }

        var remainingHand = state.Player.Hand.Where(id => id != cardId).ToList();
        var discard = new List<string>(state.Player.Discard) { cardId };

        return new GameState
        {
            Player = state.Player with
            {
                Mana = state.Player.Mana - card.Cost,
                Hand = remainingHand,
                Discard = discard
            },
            Graph = state.Graph,
            Cards = state.Cards,
            Companions = state.Companions,
            History = new List<string>(state.History) { $"played {card.Name}" },
            Battle = state.Battle,
            PendingAbilityChoices = state.PendingAbilityChoices
        };
    }

    // -------------------------------------------------------------------------
    // Helper
    // -------------------------------------------------------------------------

    private static Dictionary<string, JsonElement> JsonProperties(params (string Key, int Value)[] pairs)
    {
        var dict = new Dictionary<string, JsonElement>();
        foreach (var (key, value) in pairs)
            dict[key] = JsonSerializer.SerializeToElement(value);
        return dict;
    }
}
